[JSC] Add CallCustomAccessorGetter / CallCustomAccessorSetter DFG nodes

Constellation opened the pull request:
#### 2280a97d2334cb16aa13cf8db9a01613d99366f9
<pre>
[JSC] Add CallCustomAccessorGetter / CallCustomAccessorSetter DFG nodes
<a href="https://bugs.webkit.org/show_bug.cgi?id=263745">https://bugs.webkit.org/show_bug.cgi?id=263745</a>
<a href="https://rdar.apple.com/117551452">rdar://117551452</a>

Reviewed by Keith Miller.

When we see custom accessor getter and setter in DFG, we give up and use normal generic IC for this site.
We should not do that since

1. IC is costly. Its peak performance is fast, but setup is not free.
2. IC bloats more code.
3. Directly embedding IC&apos;s custom accessor getter / setter code into DFG / FTL is faster and more efficient.

This patch adds CallCustomAccessorGetter and CallCustomAccessorSetter. We appropriately start handling custom
accessor feedback from IC and starts putting these nodes instead of just putting GetById / PutById.

                                            ToT                     Patched

custom-accessor                        9.7385+-0.0277     ^      9.4625+-0.0374        ^ definitely 1.0292x faster
custom-accessor-materialized           9.8093+-0.0166     ^      9.5174+-0.0291        ^ definitely 1.0307x faster
custom-accessor-thin-air-setter        5.3423+-0.0150     ^      4.7375+-0.0205        ^ definitely 1.1277x faster
custom-accessor-thin-air              19.8702+-0.0815     ^     19.1854+-0.0327        ^ definitely 1.0357x faster

* JSTests/microbenchmarks/custom-accessor-thin-air-setter.js: Added.
(assert):
(test5.set const):
(test5.set for):
* Source/JavaScriptCore/bytecode/GetByStatus.cpp:
(JSC::GetByStatus::computeForStubInfoWithoutExitSiteFeedback):
* Source/JavaScriptCore/bytecode/GetByStatus.h:
* Source/JavaScriptCore/bytecode/ObjectPropertyConditionSet.cpp:
(JSC::ObjectPropertyConditionSet::isStillValid const):
* Source/JavaScriptCore/bytecode/ObjectPropertyConditionSet.h:
* Source/JavaScriptCore/bytecode/PutByStatus.cpp:
(JSC::PutByStatus::computeForStubInfo):
(JSC::PutByStatus::makesCalls const):
(JSC::PutByStatus::merge):
(JSC::PutByStatus::dump const):
* Source/JavaScriptCore/bytecode/PutByStatus.h:
* Source/JavaScriptCore/bytecode/PutByVariant.cpp:
(JSC::PutByVariant::operator=):
(JSC::PutByVariant::customSetter):
(JSC::PutByVariant::writesStructures const):
(JSC::PutByVariant::reallocatesStorage const):
(JSC::PutByVariant::makesCalls const):
(JSC::PutByVariant::attemptToMerge):
(JSC::PutByVariant::dumpInContext const):
* Source/JavaScriptCore/bytecode/PutByVariant.h:
(JSC::PutByVariant::oldStructure const):
(JSC::PutByVariant::oldStructure):
(JSC::PutByVariant::customAccessorSetter const):
(JSC::PutByVariant::domAttribute const):
* Source/JavaScriptCore/dfg/DFGAbstractInterpreterInlines.h:
(JSC::DFG::AbstractInterpreter&lt;AbstractStateType&gt;::executeEffects):
* Source/JavaScriptCore/dfg/DFGByteCodeParser.cpp:
(JSC::DFG::ByteCodeParser::check):
(JSC::DFG::ByteCodeParser::handleGetById):
(JSC::DFG::ByteCodeParser::handlePutById):
* Source/JavaScriptCore/dfg/DFGClobberize.h:
(JSC::DFG::clobberize):
* Source/JavaScriptCore/dfg/DFGDoesGC.cpp:
(JSC::DFG::doesGC):
* Source/JavaScriptCore/dfg/DFGFixupPhase.cpp:
(JSC::DFG::FixupPhase::fixupNode):
* Source/JavaScriptCore/dfg/DFGGraph.h:
* Source/JavaScriptCore/dfg/DFGMayExit.cpp:
* Source/JavaScriptCore/dfg/DFGNode.h:
(JSC::DFG::Node::hasCacheableIdentifier):
(JSC::DFG::Node::cacheableIdentifier):
(JSC::DFG::Node::hasHeapPrediction):
(JSC::DFG::Node::hasCallCustomAccessorData const):
(JSC::DFG::Node::callCustomAccessorData):
(JSC::DFG::Node::customAccessor):
* Source/JavaScriptCore/dfg/DFGNodeType.h:
* Source/JavaScriptCore/dfg/DFGPredictionPropagationPhase.cpp:
* Source/JavaScriptCore/dfg/DFGSafeToExecute.h:
(JSC::DFG::safeToExecute):
* Source/JavaScriptCore/dfg/DFGSpeculativeJIT.cpp:
* Source/JavaScriptCore/dfg/DFGSpeculativeJIT.h:
* Source/JavaScriptCore/dfg/DFGSpeculativeJIT32_64.cpp:
(JSC::DFG::SpeculativeJIT::compile):
* Source/JavaScriptCore/dfg/DFGSpeculativeJIT64.cpp:
(JSC::DFG::SpeculativeJIT::compile):
* Source/JavaScriptCore/ftl/FTLCapabilities.cpp:
(JSC::FTL::canCompile):
* Source/JavaScriptCore/ftl/FTLLowerDFGToB3.cpp:
(JSC::FTL::DFG::LowerDFGToB3::compileNode):
(JSC::FTL::DFG::LowerDFGToB3::compileCompareStrictEq):

Canonical link: <a href="https://commits.webkit.org/270028@main">https://commits.webkit.org/270028@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/2896ef04951a700351b3c405f164121589f93240

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/24324 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/2434 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/25411 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/26457 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/22389 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/4073 "Built successfully") | [  ~~🛠 mac-AS-debug~~](https://ews-build.webkit.org/#/builders/51/builds/7 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/22827 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/24568 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/47/builds/1955 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/21019 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/27047 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/1699 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/21945 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/28161 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/21165 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/22180 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/22254 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/25950 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/23622 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/1625 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/51/builds/7 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/31019 "Built successfully") | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/46/builds/2904 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/6798 "Passed tests") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/2030 "Built successfully") | | [✅ 🛠 jsc-mips](https://ews-build.webkit.org/#/builders/24/builds/30984 "Built successfully") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/3107 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/1993 "Built successfully") | | [✅ 🧪 jsc-mips-tests](https://ews-build.webkit.org/#/builders/3/builds/6474 "Passed tests") | 
<!--EWS-Status-Bubble-End-->

kmiller68 commented on Source/JavaScriptCore/dfg/DFGByteCodeParser.cpp:
```diff
@@ -4756,6 +4756,9 @@ bool ByteCodeParser::check(const ObjectPropertyCondition& condition)
     if (m_graph.watchCondition(condition))
         return true;
+
+    if (condition.kind() == PropertyCondition::Equivalence)
+        return false;
```
Why is this necessary?

Constellation replied:
The reason is that this `check` was designed to be only used for presence conditions so far.
This was true before, since only presence conditions are produced from IC and we use equivalence conditions explicitly by using `watchCondition`. However https://github.com/WebKit/WebKit/commit/667a7374f1ddff4f597489916022e7856e4e2664 changed it and it unfortunately broke our code: the current code always returns `false` from `check` if equivalent conditions are used. So these intrinsic getter `__proto__` thing no longer gets DFG nodes correctly.
This change fixes it: we use `isStillValid` instead of `structureEnsuresValidity` to accept equivalence conditions here. However, if `watchCondition` failed, we are not having a way to make it work. So, we explicitly return `false`, and this is aligned to the current behavior (currently we return `false` before reaching to this point since `structureEnsuresValidity` for equivalence condition always returns false).

kmiller68 commented on Source/JavaScriptCore/bytecode/PutByStatus.h:
```diff
@@ -53,6 +53,8 @@ class PutByStatus final {
         // It's cached as a simple store of some kind.
         Simple,
+        // It's cached for a custom accessor with a possible structure chain.
+        Custom,
```
Nit: Can we call this `CustomAccessor`?

Constellation replied:
Changed.

kmiller68 commented on Source/JavaScriptCore/bytecode/PutByVariant.cpp:
```diff
@@ -248,6 +267,35 @@ bool PutByVariant::attemptToMerge(const PutByVariant& other)
+        if (m_conditionSet.isEmpty() != other.m_conditionSet.isEmpty())
+            return false;
+
+        ObjectPropertyConditionSet mergedConditionSet;
```
Shouldn't this be `mergedConditionSet = other.m_conditionSet`?

Constellation replied:
Right, changed.

Constellation replied:
Ah, no. We have `m_conditionSet.isEmpty() != other.m_conditionSet.isEmpty()`. So if m_conditionSet is empty, then other is also empty.

kmiller68 commented on Source/JavaScriptCore/bytecode/PutByStatus.h:
```diff
@@ -111,10 +113,12 @@ class PutByStatus final {
     bool isSet() const { return m_state != NoInformation; }
     bool operator!() const { return m_state == NoInformation; }
     bool isSimple() const { return m_state == Simple; }
+    bool isCustom() const { return m_state == Custom; }
```
Nit: Can we call this `isCustomAccessor()`?

kmiller68 commented on Source/JavaScriptCore/dfg/DFGByteCodeParser.cpp:
```diff
@@ -5191,15 +5194,37 @@ void ByteCodeParser::handleGetById(
+            if (Options::useDOMJIT() && variant.domAttribute()) {
+                ASSERT(!getByStatus.makesCalls());
+                if (handleDOMJITGetter(destination, variant, base, identifierNumber, prediction)) {
+                    if (UNLIKELY(m_graph.compilation()))
```
Wow, `compilation` is a not a good function name for the Bytecode profile's compilation object...

kmiller68 reviewed (approved):
r=me with comments

Constellation committed:
[JSC] Add CallCustomAccessorGetter / CallCustomAccessorSetter DFG nodes
https://bugs.webkit.org/show_bug.cgi?id=263745
rdar://117551452

Reviewed by Keith Miller.

When we see custom accessor getter and setter in DFG, we give up and use normal generic IC for this site.
We should not do that since

1. IC is costly. Its peak performance is fast, but setup is not free.
2. IC bloats more code.
3. Directly embedding IC's custom accessor getter / setter code into DFG / FTL is faster and more efficient.

This patch adds CallCustomAccessorGetter and CallCustomAccessorSetter. We appropriately start handling custom
accessor feedback from IC and starts putting these nodes instead of just putting GetById / PutById.

                                            ToT                     Patched

custom-accessor                        9.7385+-0.0277     ^      9.4625+-0.0374        ^ definitely 1.0292x faster
custom-accessor-materialized           9.8093+-0.0166     ^      9.5174+-0.0291        ^ definitely 1.0307x faster
custom-accessor-thin-air-setter        5.3423+-0.0150     ^      4.7375+-0.0205        ^ definitely 1.1277x faster
custom-accessor-thin-air              19.8702+-0.0815     ^     19.1854+-0.0327        ^ definitely 1.0357x faster

* JSTests/microbenchmarks/custom-accessor-thin-air-setter.js: Added.
(assert):
(test5.set const):
(test5.set for):
* Source/JavaScriptCore/bytecode/GetByStatus.cpp:
(JSC::GetByStatus::computeForStubInfoWithoutExitSiteFeedback):
* Source/JavaScriptCore/bytecode/GetByStatus.h:
* Source/JavaScriptCore/bytecode/ObjectPropertyConditionSet.cpp:
(JSC::ObjectPropertyConditionSet::isStillValid const):
* Source/JavaScriptCore/bytecode/ObjectPropertyConditionSet.h:
* Source/JavaScriptCore/bytecode/PutByStatus.cpp:
(JSC::PutByStatus::computeForStubInfo):
(JSC::PutByStatus::makesCalls const):
(JSC::PutByStatus::merge):
(JSC::PutByStatus::dump const):
* Source/JavaScriptCore/bytecode/PutByStatus.h:
* Source/JavaScriptCore/bytecode/PutByVariant.cpp:
(JSC::PutByVariant::operator=):
(JSC::PutByVariant::customSetter):
(JSC::PutByVariant::writesStructures const):
(JSC::PutByVariant::reallocatesStorage const):
(JSC::PutByVariant::makesCalls const):
(JSC::PutByVariant::attemptToMerge):
(JSC::PutByVariant::dumpInContext const):
* Source/JavaScriptCore/bytecode/PutByVariant.h:
(JSC::PutByVariant::oldStructure const):
(JSC::PutByVariant::oldStructure):
(JSC::PutByVariant::customAccessorSetter const):
(JSC::PutByVariant::domAttribute const):
* Source/JavaScriptCore/dfg/DFGAbstractInterpreterInlines.h:
(JSC::DFG::AbstractInterpreter<AbstractStateType>::executeEffects):
* Source/JavaScriptCore/dfg/DFGByteCodeParser.cpp:
(JSC::DFG::ByteCodeParser::check):
(JSC::DFG::ByteCodeParser::handleGetById):
(JSC::DFG::ByteCodeParser::handlePutById):
* Source/JavaScriptCore/dfg/DFGClobberize.h:
(JSC::DFG::clobberize):
* Source/JavaScriptCore/dfg/DFGDoesGC.cpp:
(JSC::DFG::doesGC):
* Source/JavaScriptCore/dfg/DFGFixupPhase.cpp:
(JSC::DFG::FixupPhase::fixupNode):
* Source/JavaScriptCore/dfg/DFGGraph.h:
* Source/JavaScriptCore/dfg/DFGMayExit.cpp:
* Source/JavaScriptCore/dfg/DFGNode.h:
(JSC::DFG::Node::hasCacheableIdentifier):
(JSC::DFG::Node::cacheableIdentifier):
(JSC::DFG::Node::hasHeapPrediction):
(JSC::DFG::Node::hasCallCustomAccessorData const):
(JSC::DFG::Node::callCustomAccessorData):
(JSC::DFG::Node::customAccessor):
* Source/JavaScriptCore/dfg/DFGNodeType.h:
* Source/JavaScriptCore/dfg/DFGPredictionPropagationPhase.cpp:
* Source/JavaScriptCore/dfg/DFGSafeToExecute.h:
(JSC::DFG::safeToExecute):
* Source/JavaScriptCore/dfg/DFGSpeculativeJIT.cpp:
* Source/JavaScriptCore/dfg/DFGSpeculativeJIT.h:
* Source/JavaScriptCore/dfg/DFGSpeculativeJIT32_64.cpp:
(JSC::DFG::SpeculativeJIT::compile):
* Source/JavaScriptCore/dfg/DFGSpeculativeJIT64.cpp:
(JSC::DFG::SpeculativeJIT::compile):
* Source/JavaScriptCore/ftl/FTLCapabilities.cpp:
(JSC::FTL::canCompile):
* Source/JavaScriptCore/ftl/FTLLowerDFGToB3.cpp:
(JSC::FTL::DFG::LowerDFGToB3::compileNode):
(JSC::FTL::DFG::LowerDFGToB3::compileCompareStrictEq):

Canonical link: https://commits.webkit.org/270028@main
